### ISSUE TITLE
Weird distortion effect in Persona 5 

### ISSUE THREAD

**Issue by reinou** [RPCS3/rpcs3#4086]
There's a weird distortion effect affecting the whole screen when in palace or when fusing personas. I checked older revisions and it started happening after [#3772](https://github.com/RPCS3/rpcs3/pull/3772) was merged.

[Palace after #3772](https://imgur.com/LT9zPPy)
[Palace before #3772](https://imgur.com/fjAScrX)
[Persona fusion after #3772](https://imgur.com/lUSOJpW)
[Persona fusion before #3772](https://imgur.com/02A4xSS)

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1644884/RPCS3.log.gz)

**Comment by Fwha** [comment]
That effect is supposed to be there if you look at ps4 gameplay, however it seems like it's greatly exaggerated in RPCS3. Ni No Kuni had this as well in the desert area although it ended up getting fixed.

**Comment by EliEron** [comment]
I don't know if it's exaggerated as much as it is misplaced. If you look at this fusion video:
https://youtu.be/XNdF1InU80c?t=50

You can see that the effect is applied specifically to the blue part on the guillotine blades. The intensity of the effect seems more or less accurate, but the fact that it is applied to the whole screen rather than just a specific surface is clearly a bug.

**Comment by al0xf** [comment]
Indeed, this looks too distorted.

**Comment by jonwd7** [comment]
Just posting some observations.  It appears to be a screenspace effect that is simply not getting masked in some instances.  I tried to find correct examples on YouTube where the character is not moving much to minimize compression artifacts.

### Fusions
Masked to some effects as seen in the above link.

### Palace Exterior 
Seemingly **unmasked** inside the gate, so this is the only time that it looks correct on RPCS3.  I say seemingly because it varies based on camera angle.  Correct examples:
    - https://youtu.be/hrLn56pmNm8?t=2795  ... In this cutscene it looks unmasked
    - https://youtu.be/hrLn56pmNm8?t=2778  ... In this cutscene it looks completely absent.

I think it might be that there is a volume or plane in the gateway which the camera is looking through in the one angle and not in the other.

### Palace Interior (Security < 99%)
Masked to some effects(?).  On RPCS3 it does it on the entire screen (like the above video), making the palaces nearly unplayable.  Correct examples:  
   - https://youtu.be/hrLn56pmNm8?t=3264  ... Here it only appears to happen over the windows, and maybe some around the enemies.  
   - https://youtu.be/hrLn56pmNm8?t=3441 ... Here it only appears to happen over the chandelier.
   - https://youtu.be/hrLn56pmNm8?t=3958 ... Here it doesn't appear to happen anywhere on screen.
   - https://youtu.be/hrLn56pmNm8?t=5315 ... Only happens in the back left corner of this room, where there are no discernable lighting/particle effects.

### Palace Interior (Security 99%)
Unmasked while in control, partially masked during cutscenes, and a different effect from <99%. It has more high frequency zig-zag patterns.   Correct examples:
    - https://youtu.be/Rq7Rm43S7Z0?t=3215 ...  Here it's unmasked.
    - https://youtu.be/Rq7Rm43S7Z0?t=3224 ...  Here it happens mostly only over the checker floor, staircase, and candelabras.  You can see it go back to being unmasked after the dialog ends.

These observations have led me to believe that the mask may be coming from a stencil buffer made of certain geometry in the room...  So the area of distortion is volumetric, but the effect is screenspace.

**Comment by kd-11** [comment]
Its not so simple. If the mask is lost its because the heuristics could not determine what to write to that address, between depth/stencil or color. It could also have been that the write is fine, but an explicit stencil clear is lost. Due to how the hw is configured, there is no way to really know if a memory address is going to end up being used for depth/stencil or color when it is bound to both (you'd be surprised how often that happens). The fact that null also points to valid memory for render targets, the problem is compounded further; as the code gets smarter about making such decisions this will be fixed.

**Comment by jonwd7** [comment]
@kd-11 

![image](https://user-images.githubusercontent.com/170077/36067078-fc837cfa-0e83-11e8-967d-2c17339a8f18.png)

![image](https://user-images.githubusercontent.com/170077/36067082-11c0251e-0e84-11e8-86ac-14be7fdaac82.png)

I thought it'd be futile looking at the Vulkan output but I thought just maybe the stencil is already there and it's just not being properly utilized.   The stencil as seen in the above two captures is *exactly* where the distortion effect is masked.  I have double checked with original footage.  

FYI this is the throne room and hallway outside at < 99% Security, not the two clips above at 99% Security.  I verified in other clips that only the hallway floor is distorted outside the throne room, and inside the throne room the stencil is pixel perfect with where you see the distortion in that cutscene. 

As far as the other clips go, here is the chandelier one (the 2nd <99% video):

![image](https://user-images.githubusercontent.com/170077/36067256-adc41616-0e87-11e8-91bb-9e0281488c41.png)

**Comment by jonwd7** [comment]
Also I tried looking at Security 99% but it's a bit more confusing because there are two distortion effects that are out of phase with one another.  I tried to capture the 99% distortion effect but couldn't seem to find it (the stencil would be white or nearly all white).  The <99% distortion effect runs underneath it with the stencils seen above.

**Comment by jonwd7** [comment]
Removed the common functions for here:
```glsl
#version 420
#extension GL_ARB_separate_shader_objects: enable

layout(set=0, binding=19) uniform sampler2D tex1;
layout(set=0, binding=20) uniform sampler2D tex2;
layout(set=0, binding=21) uniform sampler2D tex0;
layout(std140, set = 0, binding = 2) uniform FragmentConstantsBuffer
{
	vec4 fc32;
	vec4 fc112;
	vec4 fc176;
	vec4 fc224;
	vec4 fc256;
	float fog_param0;
	float fog_param1;
	uint alpha_test;
	float alpha_ref;
	uint alpha_func;
	uint fog_mode;
	float wpos_scale;
	float wpos_bias;
	vec4 texture_parameters[16];
};

layout(location=0) in vec4 tc0;
layout(location=0) out vec4 ocol0;

void fs_main(inout vec4 h0, inout vec4 r0, inout vec4 r1)
{
	vec4 cc0;
	r1.xy = tc0.xy;
	r0.xy = (r1 * fc32.zzzz).xy;
	r0.x = texture(tex1, r0.xy * texture_parameters[1].xy).x;
	r0.x = (r0 * 2.0).x;
	r0.w = texture(tex1, tc0.xy * texture_parameters[1].xy).w;
	r0.xy = (r0.xxxx * fc112.xyxx + -fc112.xyxx).xy;
	r1.zw = (r0.xxxy * r0.wwww + r1.xxxy).zw;
	r0.z = texture2DReconstruct(tex2, tc0.xy * texture_parameters[2].xy, texture_parameters[2].z).z;
	cc0.x = vec4(equal(r0.zzzz, fc176.xxxx)).x;
	r0.xyz = texture(tex0, r1.zwzz.xy * texture_parameters[0].xy).xyz;
	h0.xyz = clamp((r0 * fc224), -65504., 65504.).xyz;
	h0.w = clamp(fc256, -65504., 65504.).w;
	if (equal(cc0.xxxx, vec4(0., 0., 0., 0.)).x) h0.xyz.x = texture(tex0, tc0.xy * texture_parameters[0].xy).xyz.x;
	if (equal(cc0.xxxx, vec4(0., 0., 0., 0.)).y) h0.xyz.y = texture(tex0, tc0.xy * texture_parameters[0].xy).xyz.y;
	if (equal(cc0.xxxx, vec4(0., 0., 0., 0.)).z) h0.xyz.z = texture(tex0, tc0.xy * texture_parameters[0].xy).xyz.z;

	if (alpha_test != 0 && !comparison_passes(h0.a, alpha_ref, alpha_func)) discard;
}

void main()
{
	vec4 h0 = vec4(0.);
	vec4 r0 = vec4(0.);
	vec4 r1 = vec4(0.);

	fs_main(h0, r0, r1);

	ocol0 = h0;
}
```

So `tex2` is the depth sampler but according to renderdoc it's swizzled RRRR:

```
 0 20: tex2 2D Image&Sampler Texture2D DSV 39635 2560x1440 D24_UNORM_S8_UINT swizzle[RRRR] 
```

I read that depending on GL version the depth and stencil actually have to be separate samplers or you have to modify the gl calls so that the stencil is available in the same sampler.  Not sure about Vulkan.

Removing
```glsl
	r0.x = texture(tex1, r0.xy * texture_parameters[1].xy).x;
	r0.x = (r0 * 2.0).x;
	r0.w = texture(tex1, tc0.xy * texture_parameters[1].xy).w;
```
Obviously gets rid of the distortion entirely.  `tex1` is a 64x64 normal map with some noise.

**Comment by kd-11** [comment]
You cant read back stencil like that, especially not on ps3. This used to work before, just get a trace from an older build and the difference should be easy to find. From previous bug reports, this game uses stencil normally with a clear+write pass and a stencil test pass so it should be simple to fix this if you have some time.

**Comment by jonwd7** [comment]
https://github.com/RPCS3/rpcs3/pull/3772/files#diff-c72ed2753dd324ed6e897f5ae10985f8R324

```
//NOTE: Memory layout is fetched as byteswapped BGRA [GBAR] (GOW collection, DS2, DeS)
//The A component (Z) is useless (should contain stencil8 or just 1)
```
I'm guessing you calling it useless might have something to do with it not being used. :)

Also, that GBAR statement doesn't even seem to be true, at least not now. The depth sampler is sent to the shader as RRRR so every channel is the same.  I verified by setting `ocol0` to the sample of `tex2` and checking each channel.  They were all the depth buffer.  The resource still clearly has the stencil according to RenderDoc so I'm not sure what you mean about PS3 not being able to do such and such, as RSX has already sent the depth buffer complete with stencil to Vulkan.  Vulkan is just not sending the stencil to the shader.

Also Z can never contain stencil8 as in the shader you do this:

```
vec4 result = decodeLinearDepth(texture(tex, coord.xy).r);
```

and then return it usually immediately. Sampling the R of a depth sampler just gives you the 24-bit depth as float obviously.

> This used to work before, just get a trace from an older build

How long ago are we talking?  I went back through all the issues and this fullscreen effect seems to have never worked until PR 3772 (as demonstrated in the OP). At least not in Vulkan. I have watched RPCS3 videos from Jul-Nov 2017 and none of them had distortion anywhere (masked or otherwise). So the shader probably wasn't getting created for this effect, so I can't very well debug older builds. 

The 99% effect *does* work on older builds, but it's not masked it turns out, so it's not really relevant to this.

**Comment by dakken21** [comment]
Not to be a bother, but I've just started trying to use rpcs3 to play Persona 5 and I noticed this messed up distortion effect, and googling led me here.  I was just wondering if there was any potential fix for this issue, even just a special version of RPCS3 with the fix.  If it's not too much hassle, I don't want to be a bother but I figured I'd ask.  Thanks!

**Comment by ruipin** [comment]
As a workaround, you can use the following in your patch.yml:

```# Game: Persona 5
# Subject: Disable "Normal" Distortion Filter (<=99% Alert & Velvet Room)
#          Workaround for RPCS3 issue #4086
# Authors: TGE, ruipin, kd-11
p5_DisableDistortion: &p5_DisableDistortion
    - [ be32, 0x00FE2E28, 0x9E001700 ]
    - [ be32, 0x00FE2E2C, 0xC801001D ]
    - [ be32, 0x00FE2E30, 0x00000000 ]
    - [ be32, 0x00FE2E34, 0x00000000 ]
    - [ be32, 0x00FE2E38, 0x1E810100 ]
    - [ be32, 0x00FE2E3C, 0xC800001D ]
    - [ be32, 0x00FE2E40, 0x00000000 ]
    - [ be32, 0x00FE2E44, 0x00000000 ]

# Game: Persona 5
# Subject: Disable "Angery" Distortion Filter (100% Alert)
# Authors: TGE, ruipin, kd-11
p5_Disable100PctDistortion: &p5_Disable100PctDistortion
    - [ be32, 0x00FE31CC, 0x9E001700 ]
    - [ be32, 0x00FE31D0, 0xC801001D ]
    - [ be32, 0x00FE31D4, 0x00000000 ]
    - [ be32, 0x00FE31D8, 0x00000000 ]
    - [ be32, 0x00FE31DC, 0x1E810100 ]
    - [ be32, 0x00FE31E0, 0xC800001D ]
    - [ be32, 0x00FE31E4, 0x00000000 ]
    - [ be32, 0x00FE31E8, 0x00000000 ]```

**Comment by dakken21** [comment]
Dude that's awesome, thank you so much! I'm about halfway through the second dungeon and while I had kind of started getting used to it, it's still very bothersome and ugly.   This totally worked, I'm sorry I had to bug you guys here, but I really couldn't find anything!!! Thanks again! YOU GUYS ROCK!!!!
